### DISCUSSION TITLE
EV Charger <> Invertor Modbus communication

### DISCUSSION THREAD

**Discussion by PatrikTrestik** [wills106/homeassistant-solax-modbus#666] [Ideas]
I started work on monitoring on the wire Modbus communication between Charger nad Inverter.
Charger FW 1.12, Inverter X3 G4 fw 1.34.

First findings:
Inverter (Master) periodicity reads EvCharger (slave 70), 3 registers from address 0x0008 - mens that Inverter is monitoring Charging Power on all phases.

Inverter also periodically writes to slave 0 (???), 10 registers from address 0x700 -  I have no idea what is it. Might be irelevant to EvCharger.

Next work: 
- connect car and check while charging
- wait for sun and check with Green mode

**Comment by wills106** [comment]
Slave 0 is a Broadcast address. So it's sending those 10 registers to anything connected to the Modbus daisy chain.

Do they match PV Generation, Grid Import / Export? I would expect the Inverter to provide that sort of Data to the Charger.

**Comment by PatrikTrestik** [comment]
> Slave 0 is a Broadcast address. So it's sending those 10 registers to anything connected to the Modbus daisy chain.
> 
> Do they match PV Generation, Grid Import / Export? I would expect the Inverter to provide that sort of Data to the Charger.

Ahh, I forget about broadcast. Great idea. Data looks like it might be some power information, but I have no sun here to be sure :)
This is what I have so far:
0, 56305, 65535, 0, 62439, 65535, 62489, 65535, 62449, 65535
?,Measured Power,?,?,Measured Power L1,?,Measured Power L2,?,Measured Power L3,?

**Comment by PatrikTrestik** [comment]
Update regarding smart functions.
"Don't charge EV from battery"
* Inverter periodically reads charging power from Charger. It probably adjusts internally not to compensate that much power from battery.
* Idea how to build it on older inverters - manipulate grid setpoint and "force" it negative to compensate charging power.

"Green charging"
* Inverter broadcasts number of information on modbus. Charger listens and can adjust according to excess PV power.
* We can write these information and Green functions should work.
* Still not fully decoded what is broadcasted

"Modbus multimaster"
* Weird message, function 255, is sent by inverter. It could be token offer to other master to start writing to modbus.
* As there is no standard for this, only small chance exists that we decode it.

**Comment by PatrikTrestik** [comment]
Inverter message decoding state:
0-?? never seen other value than 0
1-is not sum of individual L, might be excess PV power, reflected in RealtimeData API 22
2-65535 when discharging battery, 0 when charging batt/excess power
3- always zero
4,6,8 - 100% sure L1,L2,L3 grid measured power, RealtimeData API 19,20,21
5,7,9 - similar to 2 for separate L?

There is still not enough sun to do more experiments 😢

**Reply by wills106** [reply]
Thanks for documenting what you have found so far.

**Reply by PatrikTrestik** [reply]
1 - 100% excess PV power. If not charging it shows Grid Export+Battery Charge. I expect that this is all bound to PV production. If charging battery from grid, I expect that it will not report it as excess. 
When starts charging in Green mode, charger chooses power (in certain steps) which is lower than this.

2,5,7,9 - I see nothing else than "minus" indicator. When corresponding value (1,4,6,8) is negative (technically it is sInt16), this is 65535.

I think we have enough to experiment 😄 
Searching for volunteer!
- You have to have Solax EV Charger, EV Car, Inverter supported by this integration and way how to connect Ev Charger RS485 (for ex. converter).

I'm not volunteering because there is 5°C in my garage 🥶

**Reply by wills106** [reply]
> Int16
I was expecting either an Int16 or Int32 for some of the power registers. Like Grid, Battery etc as the Charger needs to know if you are charging / discharging the battery and exporting / importing from the Grid.

**Reply by PatrikTrestik** [reply]
Yes. Might be I didn't explained correctly.
In this example:
0, 56305, 65535, 0, 62439, 65535, 62489, 65535, 62449, 65535
reg1=56305=-9230 W - It was at time when my heating was on.
reg2=65535 - works like sign, I expect it is MSB of sInt32. For my PV is Int16 enough 😄 

At time there was some sun I have:
reg1=1852=1852 W - Car was disconnected and I was exporting 1022 W to grid and charging battery with roughly 800 W
reg2=0 - as expected if it is MSB
reg4/5, reg6/7, reg8/9 - exactly shows measured power to/from grid.

If I connect car in Green mode, charger selects some power level bellow reported excess power (in this example cca 1200W).
reg1 goes down to cca 600 W. Inverter used it for charging battery.

**Reply by wills106** [reply]
So the 10 registers at 0x700 are like:
![image](https://github.com/wills106/homeassistant-solax-modbus/assets/18155231/77123b77-7caa-457c-91b2-96c8ed37148c)
I wonder if 0x700 or 0x703 is use to halt or pause the Charger or perhaps change work modes?

Can you control the Charger through the Inverter? Or do you have to use the separate app / cloud instance?

**Reply by wills106** [reply]
On Single phase (At least on the Gen3) We only have `measured_power` I just tried exposing `measured_power_l1` and it just returns 0 regardless. So it isn't used, only `measured_power` exists.

![image](https://github.com/wills106/homeassistant-solax-modbus/assets/18155231/e2a1b04e-5d3b-40cc-8fe6-eed0af506208)

Would be interesting to know what happens on a Single phase charger.

Do you have to write `measured_power` only to 0x701 or is it still needed on 0x704? As they would be the same value.

**Reply by PatrikTrestik** [reply]
Reg 0x701 is not measured power.
It is excess PV power.
If there was sunny day it reports PV-House load.

**Reply by wills106** [reply]
O thanks, should be easy to implement then.

**Reply by PatrikTrestik** [reply]
Few observations how EV Charger works in Green mode. I have 3Phase charger but only 1Phase car. All this is related to single phase charging.
* Minimal charging current is 6A (not sure if in general or SolaX only)
* Charger waits to see 1380+W in 0x701. Waits for some time to see if it is stable. I don't know exactly, but if it is fluctuating around  1380W it never starts charging, requires stable 1400 for example.
* When charging in Green mode it compensates 0x701 to be 0W regulating charging power (6A - 16A in my case).
* When 0x701 drops below 1380W it stops charging. Again there is some waiting time, stop is not immediate.

If you set Green Level1 3A (this could be inacurate, observed only once)
* Charge is started at 0x701 = 690W. Remaining 690W is going from grid. Still charging 6A.
* Charge remains on 6A till 0x701 is 1380W and only than starts raising charging power to compensate.

**Reply by wills106** [reply]
> * Minimal charging current is 6A (not sure if in general or SolaX only)

I think that's industry standard. Some cars can charge at less, but others will refuse to charge at lower levels.

> * Charger waits to see 1380+W in 0x701. Waits for some time to see if it is stable. I don't know exactly, but if it is fluctuating around  1380W it never starts charging, requires stable 1400 for example.
> * When 0x701 drops below 1380W it stops charging. Again there is some waiting time, stop is not immediate.

I think on my Zappi 2.0 the wait is 30s each way by default. Can be adjusted but not recommended. I think if switched too often cars can complain and stop charging till cable disconnected and reconnected.

> * Charge is started at 0x701 = 690W. Remaining 690W is going from grid. Still charging 6A.

Do you mean it's pulling 690W from the grid? So 50% Solar and 50% from Grid.

On my Zappi, this "Green Percentage" can be adjusted. Perhaps you could manipulate 0x701 to achieve the same. Although probably not required.

I am guessing Green Mode doesn't pull from the house battery?

Can you charge from the house battery if needed? That could be chat 0x700 or 0x703 is for?

Thanks for the Investigation so far.

**Comment by PatrikTrestik** [comment]
> I think on my Zappi 2.0 the wait is 30s each way by default. Can be adjusted but not recommended. I think if switched too often cars can complain and stop charging till cable disconnected and reconnected.
Yes, It even happened with my Hyundai Ioniq Electric. It does not like if there are "clauds during green charging".  
> 
> > * Charge is started at 0x701 = 690W. Remaining 690W is going from grid. Still charging 6A.
> 
> Do you mean it's pulling 690W from the grid? So 50% Solar and 50% from Grid.
> 
> On my Zappi, this "Green Percentage" can be adjusted. Perhaps you could manipulate 0x701 to achieve the same. Although probably not required.
> 
It is in line with SolaX documents describing the 3A Green mode.

> I am guessing Green Mode doesn't pull from the house battery?
> 
> Can you charge from the house battery if needed? That could be chat 0x700 or 0x703 is for?

Not charging from battery is function of Inverter. Inverter reads EV charger(70) 3 registers from address 0x0008 (Charging Power) and stops compensating this power from battery. This function is probably not possible to simulate.
EV Charger has no knowledge if it is battery or grid.

For not smart Inverters I suggest to run charger in Green 6A and manipulate 0x701 to force 3A from battery or grid.

**Comment by ppetru** [comment]
@wills106 are you still looking at getting the Solax EV charger to work "smart" with a X3 hybrid G3 inverter? I have the same setup and wondering how I could help.

**Reply by PatrikTrestik** [reply]
I have this idea for longer time, but put aside due to lack of time.

I want to write separate integration to simulate smart functions.
* Read sensor which indicates PV surplus energy
* Write by Modbus to Charger registry for green mode (already done some investigation by spying modbus).

@ppetru are you able to help with development and/or testing?

@wills106 do you prefere "build-in" function here or dedicated repo with separate integration?

**Reply by wills106** [reply]
Not had any chance to look into this myself.

But I think it should be part of either the integration as a whole, or the plugin for the Charger.

**Comment by ppetru** [comment]
I'm happy to help with testing.

After writing the previous reply I discovered that EVCC should be able to solve my issue (to be confirmed, still waiting for a modbus to ethernet converter). If it doesn't, then I could also help with development if needed.

**Reply by PatrikTrestik** [reply]
I'm working on proxy integration which will simulate Gree/Smart functions.
See here: https://github.com/PatrikTrestik/homeassistant-solax-evc-proxy

**Comment by PatrikTrestik** [comment]
If someone wants to test, here is first beta: https://github.com/PatrikTrestik/homeassistant-solax-evc-proxy

Please see README, post issues directly to repo, not here.
